opm/autodiff/VFPHelpers.hpp

```diff
@@ -0,0 +1,970 @@
+    double factor_; // Interpolation factor
+};
+
+
+
+
+
+
+/**
+ * Helper function to find indices etc. for linear interpolation
+ */
+inline InterpData findInterpData(const double& value, const std::vector<double>& values) {
+    InterpData retval;
+
+    //If we only have one value in our vector, return that
+    if (values.size() == 1) {
+        retval.ind_[0] = 0;
+        retval.ind_[1] = 0;
+        retval.inv_dist_ = 0.0;
+        retval.factor_ = 0.0;
+    }
+    // Else search in the vector
+    else {
+        //First element greater than or equal to value
+        //Start with the second element, so that floor_iter does not go out of range
+        //Don't access out-of-range, therefore values.end()-1
+        auto ceil_iter = std::lower_bound(values.begin()+1, values.end()-1, value);
+
+        //Find last element smaller than range
+        auto floor_iter = ceil_iter-1;
+
+        //Find the indices
+        retval.ind_[0] = floor_iter - values.begin();
+        retval.ind_[1] = ceil_iter - values.begin();
+
+        //Find interpolation ratio
+        double dist = (*ceil_iter - *floor_iter);
+        if (std::abs(dist) > 0.0) {
+            //Possible source for floating point error here if value and floor are large,
+            //but very close to each other
+            retval.inv_dist_ = 1.0 / dist;
+            retval.factor_ = (value-*floor_iter) * retval.inv_dist_;
+        }
+        else {
+            retval.inv_dist_ = 0.0;
+            retval.factor_ = 0.0;
+        }
+    }
+
+    return retval;
+}
+
+
+
+
+
+
+/**
+ * An "ADB-like" structure with a single value and a set of derivatives
+ */
+struct adb_like {
+    adb_like() : value(0.0), dthp(0.0), dwfr(0.0), dgfr(0.0), dalq(0.0), dflo(0.0) {};
+    double value;
+    double dthp;
+    double dwfr;
+    double dgfr;
+    double dalq;
+    double dflo;
+};
+
+inline adb_like operator+(
+        adb_like lhs,
+        const adb_like& rhs) {
+    lhs.value += rhs.value;
+    lhs.dthp += rhs.dthp;
+    lhs.dwfr += rhs.dwfr;
+    lhs.dgfr += rhs.dgfr;
+    lhs.dalq += rhs.dalq;
+    lhs.dflo += rhs.dflo;
+    return lhs;
+}
+
+inline adb_like operator-(
+        adb_like lhs,
+        const adb_like& rhs) {
+    lhs.value -= rhs.value;
+    lhs.dthp -= rhs.dthp;
+    lhs.dwfr -= rhs.dwfr;
+    lhs.dgfr -= rhs.dgfr;
+    lhs.dalq -= rhs.dalq;
+    lhs.dflo -= rhs.dflo;
+    return lhs;
+}
+
+inline adb_like operator*(
+        double lhs,
+        const adb_like& rhs) {
+    adb_like retval;
+    retval.value = rhs.value * lhs;
+    retval.dthp = rhs.dthp * lhs;
+    retval.dwfr = rhs.dwfr * lhs;
+    retval.dgfr = rhs.dgfr * lhs;
+    retval.dalq = rhs.dalq * lhs;
+    retval.dflo = rhs.dflo * lhs;
+    return retval;
+}
+
+
+
+
+
+
+/**
+ * Helper function which interpolates data using the indices etc. given in the inputs.
+ */
+#ifdef __GNUC__
+#pragma GCC push_options
+#pragma GCC optimize ("unroll-loops")
```

Comment:
Tested in other code previously with good effect. Have not tested in this code / compiler combination, so may have zero effect, but should not have a negative effect at least.